Add importer for numeric initial data

nilsvu opened the pull request:
## Proposed changes

This is a parallel component and associated actions that allow loading a volume data file from disk and distribute its tensor data to elements. It can be used to set numeric initial data for evolutions or a numeric initial guess for elliptic solves.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add SetData simple action

Takes a tagged tuple and mutates the DataBox with its content

geoffrey4444 commented on src/Parallel/ArrayIndex.hpp:
```diff
@@ -44,13 +44,22 @@ struct ArrayIndex : public CkArrayIndex {
+      : CkArrayIndex(array_index),
+        array_index_(reinterpret_cast<Index*>(CkArrayIndex::data())) {
+    ASSERT(CkArrayIndex::nInts == sizeof(Index) / sizeof(int),
+           "Got unexpected nInts.");
```
Can you please make this error message more clear?

geoffrey4444 commented on src/DataStructures/DataBox/Actions/SetData.hpp:
```diff
@@ -0,0 +1,48 @@
+  template <typename... DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent,
+            Requires<sizeof...(DbTags) != 0> = nullptr>
```
Please explain, just here on GitHub so I can learn, how this TMP Requires<> = nullptr works?

kidder replied:
@geoffrey4444  A template with parameter `Requires<CONDITION> = nullptr` will only compile if `CONDITION` evaluates to `true`, and will produce a compiler error if `CONDITION` is `false`. 
It works by transforming `Requires<CONDITION>` to the type `std::nullptr_t` if the CONDITION is `true` and to a bogus type if CONDITION is `false`.   The `= nullptr` is a default template parameter so that you don't need to specify a parameter when using the template.    See the documentation of `Utilities/Requires.hpp` in the `Utilities` topic for more details

nilsdeppe replied:
@kidder asked me to "fact check" this in person. He's very close. A `false` condition removes a function from overload resolution so the compiler is not able to use that function. What this means is you can disable a function from being used based on a certain condition, and have a separate one be used if the condition is false. You get a compiler error if *all* functions with that name were not usable because of type mismatches or `Requires` failures. The compiler error actually just needs to be `no such function found`, but since that would make it nearly impossible to figure out what is happening the shepherds of dragons (compiler devs) have taken mercy on us and provided lists of functions with matching names and why they couldn't be used.

geoffrey4444 reviewed (commented):
I took a quick look and had a couple of questions, but since this is going to change a lot with PDALs, @nilsleiffischer please assign me to review once it's no longer a Draft PR

nilsvu commented:
@geoffrey4444 @nilsdeppe I have rebased this, added tests and cleaned up everything. It's ready for review. Could you take a look when you get the chance?

geoffrey4444 commented:
I will! Traveling today but will try as soon as I can.

Geoffrey 

Sent from my iPhone

> On Aug 9, 2019, at 09:27, Nils Leif Fischer <notifications@github.com> wrote:
> 
> @geoffrey4444 @nilsdeppe I have rebased this, added tests and cleaned up everything. It's ready for review. Could you take a look when you get the chance?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

nilsvu commented:
@geoffrey4444 Rebased on latest `develop` (in particular the new option handling).

kidder commented:
@geoffrey4444 please look at

geoffrey4444 commented:
Hi @nilsdeppe looks like there are conflicts and Travis errors...would you please fix? I'm starting to review this, and that's the first thing that jumps out at me

kidder commented:
@geoffrey4444 I assume you meant your comment for @nilsleiffischer

nilsdeppe commented:
I assume you mean @nilsleiffischer, not me, @geoffrey4444 :)

nilsvu commented:
@geoffrey4444 yep, rebased this again. The commit _"Add utility function to VolumeData"_ is added in #1722, so you don't need to look at that one.

geoffrey4444 commented:
Thanks @nilsleiffischer! There are still conflicts and compile failures, though

geoffrey4444 commented on src/IO/Observer/ArrayComponentId.cpp:
```diff
@@ -21,4 +21,9 @@ bool operator!=(const ArrayComponentId& lhs,
+
+std::ostream& operator<<(std::ostream& os,
+                         const ArrayComponentId& array_component_id) noexcept {
+  return os << std::hash<ArrayComponentId>{}(array_component_id);
```
So I understand, why the {} before (array_component_id)?

nilsvu replied:
`std::hash` is a struct, so we need to instantiate it to call it: https://en.cppreference.com/w/cpp/utility/hash

nilsdeppe commented:
Since every build has failed to compile I'm assuming it won't work locally. @nilsleiffischer please don't ask for reviews on things that don't build. It's not unreasonable to make sure that the code builds and tests pass on one compiler locally before pushing. It wastes reviewers time to have to look at or think about PRs that don't compile, not to mention that it also wastes an enormous amount of computing resources.

nilsvu commented on src/Parallel/ArrayIndex.hpp:
```diff
@@ -44,13 +44,25 @@ struct ArrayIndex : public CkArrayIndex {
+  ArrayIndex(const CkArrayIndex& array_index)  // NOLINT
+      : CkArrayIndex(array_index),
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        array_index_(reinterpret_cast<Index*>(CkArrayIndex::data())) {
```
Add to class documentation:

- Charm++ allocates CkArrayIndex memory, can be configured to some size (3 ints is standard)
- We place the `Index` into this buffer using placement `new`
- `CkArrayIndex::data()` can be safely reinterpreted as an `Index*`

nilsvu commented on src/ParallelAlgorithms/Actions/SetData.hpp:
```diff
@@ -0,0 +1,53 @@
+template <typename TagsList>
+struct SetData;
+
+template <typename... Tags>
```
cond this

nilsvu commented on src/ParallelAlgorithms/Actions/SetData.hpp:
```diff
@@ -0,0 +1,53 @@
+/*!
+ * \ingroup ActionsGroup
+ * \brief Mutate the DataBox tags in `TagsList` according to the `data`.
+ *
```
give example use case

nilsvu commented on tests/Unit/IO/DataImporter/Test_Tags.cpp:
```diff
@@ -0,0 +1,14 @@
+#include "IO/DataImporter/Tags.hpp"
+
+SPECTRE_TEST_CASE("Unit.IO.DataImporter.Tags", "[Unit][IO]") {
+  CHECK(db::tag_name<importer::Tags::RegisteredElements>() ==
```
test the option tags

nilsvu commented on src/IO/DataImporter/ElementActions.hpp:
```diff
@@ -0,0 +1,55 @@
+namespace importer {
+namespace Actions {
+
+struct RegisterElementWithSelf;
```
cond this

nilsvu commented on src/IO/DataImporter/DataFileReaderActions.hpp:
```diff
@@ -0,0 +1,219 @@
+               ? ""
+               : "_" + tensor.component_name(
+                           tensor.get_tensor_index(component_index));
+  }
```
move into `Tensor`

nilsvu commented on src/IO/DataImporter/DataFileReaderActions.hpp:
```diff
@@ -0,0 +1,219 @@
+               : "_" + tensor.component_name(
+                           tensor.get_tensor_index(component_index));
+  }
+  static size_t get_observation_id(const h5::VolumeData& volume_file,
```
move into `VolumeData`

nilsvu reviewed (commented):
review with @nilsdeppe

nilsvu commented:
Factored out a bunch of small changes into #1772.

@nilsdeppe please take a look at the fixups.

nilsvu commented:
@geoffrey4444 @nilsdeppe I rebased this on #1772 that added all the utility functions, so this PR is now reduced to the parallel components and actions. There are still three fixup commits for @nilsdeppe, and @geoffrey4444 please review when you get the chance. Cheers!

geoffrey4444 commented on src/Elliptic/Executables/Poisson/SolvePoissonProblem.hpp:
```diff
@@ -71,6 +74,13 @@ struct Metavariables {
+  // setting up the executable instead.
+  using initial_guess = tmpl::type_from<analytic_solution_tag>;
+  // using initial_guess = elliptic::NumericInitialGuess<system>;
+
```
Please merge #1812 and fix this before merging so that you actually build and test numerical import in an executable

geoffrey4444 reviewed (changes requested):
Please implement #1812 before merging...then I'm happy

nilsvu commented:
@geoffrey4444 @nilsdeppe I removed the commit that adds the numeric initial guess to the elliptic executable to make this PR independent of #1812.

@nilsdeppe could you take a look at the fixups please?

kidder commented:
@nilsdeppe @geoffrey4444 please look at

geoffrey4444 commented on src/ParallelAlgorithms/Actions/SetData.hpp:
```diff
@@ -0,0 +1,53 @@
+
+template <typename... Tags>
+struct SetData<tmpl::list<Tags...>> {
+  template <
```
Nitpick, but did you clang format? Normally I thought clang-format doesn't leave spaces like this unless the <> aren't balanced

nilsvu replied:
yea, this is what clang-format wants :)

geoffrey4444 commented on src/ParallelAlgorithms/Actions/SetData.hpp:
```diff
@@ -0,0 +1,53 @@
+      typename ParallelComponent, typename DataBox, typename Metavariables,
+      typename ArrayIndex,
+      Requires<cpp17::conjunction_v<db::tag_is_retrievable<Tags, DataBox>...>> =
+          nullptr>
```
So I can learn, what is the nullptr doing? I get that the overall effect is to require that all tags are retrievable, just wondering why you assign the Requires<> type to nullptr. Also wondering why Requires<> isn't taking a type (e.g. cpp17::conjunction_t)?

nilsvu replied:
Requires is our replacement for std::enable_if, please take a look at the docs: https://spectre-code.org/group__utilitiesgroup#ga084d535585c18d4e80f8ec6052d341a3

I suppose it needs a default value so that the template parameter doesn’t need to be specified when instantiating the template. It’s nullptr because Requires is of type std::nullptr_t if I understand correctly. It could probably also be implemented taking a std::bool_constant type instead of the bool, perhaps taking a bool is easier on compile times.. @nilsdeppe can confirm or correct me here

geoffrey4444 commented on src/ParallelAlgorithms/Actions/SetData.hpp:
```diff
@@ -0,0 +1,53 @@
+  static void apply(DataBox& box,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    tuples::TaggedTuple<Tags...> data) noexcept {
```
Is data not const so that we can std::move its contents instead of copying them?

nilsvu replied:
Yes! (I hope this works as expected)

geoffrey4444 commented on tests/Unit/ParallelAlgorithms/Actions/Test_SetData.cpp:
```diff
@@ -0,0 +1,50 @@
+
+namespace {
+
+struct SomeNumber : db::SimpleTag {
```
SomeInt?

nilsvu replied:
It’s a tag used only for this test and I don’t think this name change makes the test any clearer, so I’ll keep it as-is unless you feel strongly about it for some reason :)

geoffrey4444 commented on src/IO/DataImporter/DataFileReader.hpp:
```diff
@@ -0,0 +1,76 @@
+ *
+ * Each element of the array parallel component must register itself before
+ * data can be sent to it. To do so, invoke
+ * `importer::Actions::RegisterWithImporter` on each the element. In a
```
on each element

geoffrey4444 commented on tests/Unit/ParallelAlgorithms/Actions/Test_SetData.cpp:
```diff
@@ -0,0 +1,50 @@
+};
+
+struct Metavariables {
+  using component_list = tmpl::list<Component<Metavariables>>;
```
Another newbie question. How is this not infinite recursion, putting the name of the struct containing component_list into component_list? I'd love to better understand how this line works

nilsvu replied:
If I understand correctly, type aliases are lazily instantiated, so only when `::component_list` is retrieved will the `Component<Metavariables>` be instantiated. I think this means we need to be careful to avoid recursion when using the alias within the class (only if it’s a class template?), but using it from outside the class is fine.. again, e.g. @nilsdeppe or @wthrowe can confirm or correct me here.

geoffrey4444 commented on src/IO/DataImporter/DataFileReaderActions.hpp:
```diff
@@ -0,0 +1,219 @@
+ * `importer::DataFileReader` component.
+ */
+struct RegisterElementWithSelf {
+  template <
```
Is this spacing the output of clang-format?

nilsvu replied:
yes, this is clang-formatted.

geoffrey4444 commented on src/IO/DataImporter/DataFileReaderActions.hpp:
```diff
@@ -0,0 +1,219 @@
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const observers::ArrayComponentId& array_component_id,
+                    const std::string& grid_name) noexcept {
```
Newbie question: so grid_name is the name of a specific element? Are you storing their names as strings because you'll need these strings when reading volume data from an h5 file?

nilsvu replied:
Yes, the `grid_name` is the string that identifies the element in the H5 file. We currently use the ElementId for this in the observers, so the importer does the same.

geoffrey4444 commented on tests/Unit/IO/DataImporter/Test_Tags.cpp:
```diff
@@ -7,8 +7,42 @@
+struct ExampleDataGroup {
+  using group = importer::OptionTags::Group;
+  static std::string name() noexcept { return "ExampleData"; }
+  static constexpr OptionString help = "halp";
```
This help string seems useless to me...isn't "Example help string" a little better?

nilsvu replied:
This string is used all over the testing code for help strings that are completely irrelevant, probably for no other reason than that someone started doing it and others copied. I have to fix a typo elsewhere so I’ll also change this.

nilsdeppe replied:
This started as a hat tip to a former Cornell grad student, FYI :)

geoffrey4444 reviewed (dismissed):
Fixups look ok to me...I just have a bunch of newbie questions...when you get a chance, if you could answer them, it would help me better understand how some parts of this PR are working :)

nilsvu commented:
@nilsdeppe could you take a look at this PR and check the fixups?

nilsdeppe commented:
Fixups LGTM. I'm assuming you didn't change anything else with the code,right?

nilsvu commented:
@nilsdeppe yes I believe so. When rebasing on #1772 I just dropped the commits that were already in develop.

Squashed the fixups (including the typo @geoffrey4444 mentioned above) and rebased. @geoffrey4444 @nilsdeppe please take a look, and thanks for reviewing!

Edit: @geoffrey4444 was faster :)

nilsdeppe commented on src/IO/DataImporter/DataFileReader.hpp:
```diff
@@ -0,0 +1,76 @@
+#include "Parallel/PhaseDependentActionList.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+
+namespace importer {
```
How about `initial_data_importer`? `importer` seems rather vague in my opinion

nilsvu replied:
It's initial data for evolutions and an initial guess for elliptic solves.. the component is quite general; it can read a volume data file and populate tensors in each element's DataBox from it. So how about `DataImporter` (that's currently the directory name), or perhaps you have other suggestions?

nilsdeppe replied:
I don't love `DataImporter` since that's literally anything that reads things from anywhere. How is the initial guess not initial data for an elliptic solve?

nilsvu replied:
Could be misleading to do an (elliptic) initial data solve and import „initial data“ for the initial guess, don‘t you think?

What’s the inverse to `Observers`? How about `TensorDataImporter` or `VolumeDataImporter`? A shorter name would be better..
Or we place it in the `observers` namespace since it’s observers in reverse?

nilsdeppe replied:
I don't think it's confusing. I think the issue people are confused about is that people think elliptic problems are only for initial data :roll_eyes: 

It should definitely not be in observers namespace since it's not writing data to disk.

nilsvu replied:
So we have two options that may be mildly confusing in different ways:
- `importer` doesn’t make clear it’s for volume tensor data
- `initial_data_importer` may be confusing when solving for initial data

I’m actually not quite sold on `initial_data_importer`.. it’s very long, we don’t have any other namespace that’s this long or has underscores, instead we commonly use abbreviations for namespaces. Also, the importer is not really limited to initial data; it can load tensor data at any time.

How about `importer`, `import`, `importing` or `imprt`, and I name the parallel component `VolumeDataFileReader` instead? The we can use the namespace for other possible data importing components that aren’t tensor volume data, and in `IO` we have a namespace for “in” and one for “out” :)

kidder replied:
I don't like `initial_data_importer` because there is nothing that ties this to initial data.  It literally just reads in data the has been written in the `h5::VolumeData` format.  Therefore in theory it could be used as the reader for something like SpEC's ApplyObservers.  

The new namespace is to be the reading equivalent of `observers`
The new component is to be the reading equivalent of `observers::ObserverWriter`, but specifically for volume data.
The new action is to be the reading equivalent of `observers::ThreadedActions::WriteVolumeData`

I would suggest WriteVolumeData -> ReadVolumeData,  ObserverWriter -> VolumeDataReader, and I can't think of a better name than `importers` for the namespace...

nilsdeppe replied:
If my understanding of the code is correct it does not just read in data, it assumes that there is an element array component. I suspect we will find it rather limited in its utility

nilsdeppe replied:
Here's why I don't think this code should be viewed as anything more general than reading initial data:

1. As far as I can tell it requires the exact element structure to present in the file and so does not do any interpolation.
2. It requires creating lots of elements in a parallel component in order to read the data to begin with.
3. For ApplyObservers style transformations we almost certainly don't want to create each element and split the data, we would want to operate on the collection of data in the file. If the data is too large for a single node we would want a more efficient splitting than each element.
4. There's no infrastructure for creating elements based on the contents of the file as far as I can tell. This restricts the code even in terms of reading in initial data.

nilsvu replied:
Yes, the current implementation simply “observes in reverse”. We’ll probably add features such as interpolations once we need them. So how about we do such things in an `importers` namespace, like we have components for reductions and volume data in `observers`?

nilsdeppe replied:
Sure, I'd like to sketch out the extra features we want to have and then file an issue as part of this PR. Would that be okay?

What I think we want (in no particular order):

- Support for interpolating to an arbitrary grid.
- Support for extrapolation to grid points if desired (BBH initial data used to do this in SpEC, I'm not sure it still does)
- The option to retrieve the data across all elements of a particular tensor or tensor component stored in a particular file. (I'm not sure how much this can be done in the same framework, might just need an extra action).
- Create elements of a DG element array from the elements in the file(s). This will be useful for getting an initial AMR hierarchy for the evolution code from the elliptic solver AMR grid used while solving for initial data.

Anything else?

nilsleiffischer replied:
Sounds good to me. #1926

nilsvu replied:
Sounds good to me. #1926

nilsdeppe commented on tests/Unit/IO/CMakeLists.txt:
```diff
@@ -17,6 +17,8 @@ set(LIBRARY_SOURCES
   Test_VolumeData.cpp
   )
 
+add_subdirectory(DataImporter)
```
Why are you adding this in the middle? It's creating it's own library so add it at the front or the end.

nilsdeppe reviewed (changes requested):
The main thing this is missing is documentation on how to use it. Another thing is that the `importer` namespace is not in any group. I'm not sure if there is a good group for it, but having a tutorial on how to use this would at least offset that part of the issue.

kidder commented on src/IO/DataImporter/DataFileReaderActions.hpp:
```diff
@@ -0,0 +1,197 @@
+      for (auto& element_and_name : get<Tags::RegisteredElements>(box)) {
+        const CkArrayIndex& raw_element_index =
+            element_and_name.first.array_index();
+        // Check if the parallel component of the registered element matches the
```
I don't understand; why would multiple components register their elements?

nilsvu replied:
Every element from any component that will receive data at some point must register (same for observers). Then the importer can be invoked any number of times with any file name or observation ID, and will distribute the data to the invoking component’s elements.

In other words, since the importer doesn’t know ahead of time which component will invoke it, the elements must uniquely identify themselves with their ID and their component.

kidder replied:
ah okay; please augment your test to show a case where multiple components read data

nilsvu replied:
done, please take a look at the fixup

kidder commented:
Also have you checked that you get useful error messages if you try to import non-existent data, incorrect extents, non-existent mesh, incorrect observation values, incorrect file name, etc.  I expect most of these to trigger errors from existing `IO` code, but you should double check this...

nilsvu commented:
- Performed the renames we discussed to mirror the `observers`, i.e. `WriteVolumeData` -> `ReadVolumeData`, `ObserverWriter` -> `VolumeDataReader` and reflecting the intention that we are going to add more to the `importers` namespace that may not be related to volume data.
- Also extended the tests so they have a single volume data reader importing data for multiple components from either the same or different files, as requested by @kidder. Please take a look at the fixups.
- The `IO` code is indeed responsible for most of the error messages, e.g. when trying to read ill-formed data. Checking consistency of the loaded data with e.g. the mesh is up to the receiver of the data, i.e. the callback action on the elements. The `SetData` action just writes the data to the DataBox, but we can (and probably should) add callback actions that do validation checks in the future.
- I'll also add more documentation ASAP.

nilsvu committed:
Add dev guide for importing data

nilsvu commented:
Also added the documentation now. Please take a look @kidder @nilsdeppe @geoffrey4444.

kidder commented:
I am fine with the fixups and the documentation

geoffrey4444 commented:
I'm also happy with the fixups and docs

nilsvu commented:
@nilsdeppe could you take a look at the fixups and the docs?

nilsvu commented:
@nilsdeppe please take a look.

nilsvu commented:
@nilsdeppe could you take a look

nilsdeppe commented on tests/Unit/IO/Importers/Test_VolumeDataReaderAlgorithm.hpp:
```diff
@@ -50,28 +50,100 @@ struct VectorFieldTag : db::SimpleTag {
+enum class Grid { Fine, Coarse };
+
+template <Grid TheGrid>
+std::string grid_name() noexcept {
```
Why not make this an `operator<<`?

nilsdeppe commented on docs/DevGuide/ImportingData.md:
```diff
@@ -0,0 +1,60 @@
+
+To register elements of your array parallel component for receiving volume data,
+invoke the `importers::Actions::RegisterWithVolumeDataReader` action in the
+`phase_dependent_action_list`. Here's an example from the `ElementArray`
```
This sentence seems odd since all the snippet shows is a PhaseActions, not a component

nilsvu replied:
> Here's an example from the `ElementArray` component mentioned in the previous code snippet

What is unclear to you? With "from" I mean that the following snippet is code within the `ElementArray` class.

I removed the sentence in a fixup, perhaps that removes confusion

nilsdeppe commented on docs/DevGuide/ImportingData.md:
```diff
@@ -0,0 +1,60 @@
+  write a new action that does a few consistency checks on the data before
+  moving it into the DataBox.
+
+Once you have assembled the `importers::ThreadedActions::ReadVolumeData` action
```
I don't understand what you mean by "assemble". What's not clear to me from this documentation is how to actually get the data loaded onto the ElementArray. The registration part is clear, but then for the phase `ImportData` I don't know what to do.

nilsvu replied:
This is written earlier in the docs:

> Here's an example from the ElementArray component that assembles such an action:
>
> ```cpp
> using import_fields = tmpl::list<ScalarFieldTag, VectorFieldTag<Dim>>;
> using read_element_data_action = importers::ThreadedActions::ReadVolumeData<
>      VolumeDataOptions<TheGrid>, import_fields,
>      ::Actions::SetData<import_fields>, ElementArray>;
> ```

Then I explain some of the template parameters, and then:

> Once you have assembled the importers::ThreadedActions::ReadVolumeData action you can invoke it like this:
>
> ```c++
>  static void execute_next_phase(
>       const typename Metavariables::Phase next_phase,
>       Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache) noexcept {
>     auto& local_cache = *(global_cache.ckLocalBranch());
>     Parallel::get_parallel_component<ElementArray>(local_cache)
>         .start_phase(next_phase);
>     if (next_phase == Metavariables::Phase::ImportData) {
>       Parallel::threaded_action<read_element_data_action>(
>           Parallel::get_parallel_component<
>               importers::VolumeDataReader<Metavariables>>(local_cache));
>     }
>   }
> ```

So with "assemble" I mean instantiate the `ReadVolumeData` action template with a bunch of carefully chosen template parameters. And I thought the snippet makes it quite clear how to invoke the action.

I added another paragraph which hopefully gives a better overview of what's going on.

nilsdeppe replied:
Oh, I see now what I don't understand. You should combine the `using import_fields`... snippet and the last snippet into one, then walk through what it does. The giant two bullets in the middle makes it difficult to associate the code. You could also leave the snippets separate and put the last one first, explaining how to invoke the volume read, then talking about what the action does. I think there is way too much info between the random type aliases that I thought were some special things the actions search for in the metavariables and the actual invokation could.

nilsvu replied:
I see the confusion. Reformulated as you suggested in the last fixup commit. Is it clearer now?

nilsdeppe commented on docs/DevGuide/ImportingData.md:
```diff
@@ -0,0 +1,60 @@
+\cond NEVER
```
This should be in `docs/Tutorials`

nilsvu replied:
Isn't `Tutorials` for users of our (compiled) code and `DevGuide` for developers? I think this should be where "Creating executables" is because it is about adding importing functionality to an executable. Does that make sense?

nilsdeppe replied:
No, this is code that can easily adjust the compile-time input file (metavariables), it does not go over how to write a new importer. "Creating executables" is writing a new executable from scratch, not enabling a feature we provide.

nilsvu replied:
Hm, I don't expect that users of executables will have to worry about this code much. I'll add support for importing numeric initial data to the DgElementArray and the evolution executables once this PR is merged. Then `NumericInitialData` will be treated just like we currently handle analytic solutions and data: CMake passes it as the `InitialData` template parameter to the Metavariables. So the user won't need to change anything by hand.
Once executables support numeric initial data, how about we add another user tutorial that describes how to get data into the spectre volume data format so it can be used as initial data?

nilsdeppe replied:
Just so understand, you are envisioning some more global way of enable switching between numerical and analytic initial data? What do you think that will look like?

nilsdeppe replied:
@kidder any thoughts on this?

nilsvu replied:
Here's the code for the GH executable that I intend to submit once this PR is merged: https://github.com/nilsleiffischer/spectre/commit/810c9064ae9fc064e65ee8aaa49a9df38e950b73
The link is to the last commit on my [import_num_initial_data_in_gh](https://github.com/nilsleiffischer/spectre/commits/import_num_initial_data_in_gh) branch. It's probably easier to discuss this in a call :)

nilsvu replied:
Brief overview: The evolution Metavariables already take `InitialData` as a template parameter. It's currently either an analytic solution or analytic data class, but now it can also be `NumericInitialData`. A bunch of compile-time switches (like the ones we already have for analytic solution/data) determine if the importer parallel component should be part of the component_list, for example. The DgElementArray component is capable of importing numeric initial data and does so if requested.

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the existing fixups

nilsvu commented:
done @nilsdeppe, please take a look.

nilsdeppe commented on docs/DevGuide/ImportingData.md:
```diff
@@ -0,0 +1,64 @@
+\snippet Test_VolumeDataReaderAlgorithm.hpp register_action
+
+Now you're all set to load a data file. To do so, invoke the
+`importers::ThreadedActions::ReadVolumeData` action on the
```
... action from the `execute_next_phase` function of the `ElementArray` component on the...

nilsdeppe commented on docs/DevGuide/ImportingData.md:
```diff
@@ -0,0 +1,64 @@
+`importers::ThreadedActions::ReadVolumeData` action on the
+`importers::VolumeDataReader`:
+
+\snippet Test_VolumeDataReaderAlgorithm.hpp read_data_action
```
I think the snippet would be easier to parse with a blank line between the two type aliases. What do you think?

nilsdeppe commented on docs/DevGuide/ImportingData.md:
```diff
@@ -0,0 +1,64 @@
+
+\snippet Test_VolumeDataReaderAlgorithm.hpp invoke_readvoldata
+
+The example shows the `execute_next_phase` function of an array parallel
```
How about something like:

In the first snippet the `read_element_data_action` type alias is set to a specific specialization of the generic `importers::ThreadedActions::ReadVolumeData`. The first template parameter to `ReadVolumeData` is the option tag (or is this just something to mark the group??) from the input file (see the first bullet below for more detail), the second template parameter is a typelist of the tags to import and fill with the volume data, the third template parameter is the action used to set/fill the data (usually `Actions::SetData`, see the second bullet below for more detail), and the last template parameter is the component on which `Actions::SetData` will be invoked to set the imported data. The `ReadVolumeData`/`read_element_data_action` should be invoked from the `execute_next_phase` function as in the second snippet just above.

nilsdeppe reviewed (commented):
I think the changes you've made are a good improvement. Here's some more suggestions that I think would help clarify further what is going on :)

nilsdeppe commented:
I'm happy for you to squash these dox fixups immediately, including any new ones :)

nilsvu commented:
@nilsdeppe done and squashed. please take a look :)

kidder commented:
fails to compile...

nilsvu commented:
Rebased on develop

kidder commented on src/ParallelAlgorithms/Actions/SetData.hpp:
```diff
@@ -22,7 +22,7 @@ namespace Actions {
  *
  * An example use case for this action is as the callback for the
- * `importer::ThreadedActions::ReadElementData`.
+ * `importers::ThreadedActions::ReadVolumeData`.
```
this should be in the SetData commit

nilsvu replied:
sorry, fixed.

nilsvu committed:
Add importer for numeric volume data

This is a parallel component and associated actions that allow
loading a volume data file from disk and distribute its tensor data
to elements. It can be used to set numeric initial data for
evolutions or a numeric initial guess for elliptic solves.

nilsvu commented:
@kidder @geoffrey4444 @nilsdeppe looks like this PR is good to go. Could you take a final look? Thanks for reviewing everyone :)

kidder commented:
@geoffrey4444 all this needs is your final approval

kidder commented:
@geoffrey4444 please file an issue if you were not happy with this

geoffrey4444 commented:
I would have approved but was tied up in search committee stuff...I'm very happy that this is merged!
